scio-extra/src/main/scala/com/spotify/scio/extra/sparkey/package.scala

```diff
@@ -224,48 +226,75 @@ package object sparkey extends SparkeyReaderInstances {
+      val basePath = if (path == null) tempPath else path
+      val nonShardedUri = SparkeyUri(basePath, self.context.options)
+      require(!nonShardedUri.exists, s"Sparkey URI $nonShardedUri already exists")
+      val uri = ShardedSparkeyUri(basePath, self.context.options)
+      require(!uri.exists, s"Sparkey URI $uri already exists")
+
+      logger.info(s"Saving as Sparkey with $numShards shards: $basePath")
+
+      val uriCoder = CoderMaterializer.beam(self.context, Coder[JList[SparkeyUri]])
+
+      self.transform { collection =>
+        val shards = collection
+          .groupBy { case (k, _) => floorMod(w.shardHash(k), numShards).toShort }
+          .map {
+            case (shard, xs) =>
+              shard -> writeToSparkey(
+                uri.sparkeyUriForShard(shard, numShards),
+                maxMemoryUsage,
+                xs.asJava
+              )
+          }
+
+        val shardsMap = shards.asMapSideInput
+
+        val uris = shards.context
+          .parallelize((0 until numShards).map(_.toShort))
+          .withSideInputs(shardsMap)
+          .map {
+            case (shard, sideContext) =>
+              sideContext(shardsMap).getOrElse(
+                shard,
+                writeToSparkey(
+                  uri.sparkeyUriForShard(shard, numShards),
+                  maxMemoryUsage,
+                  Iterable.empty[(K, V)].asJava
+                )
+              )
+          }
+          .toSCollection
+
+        uris.context
+          .wrap {
+            val view = uris.applyInternal(View.asList())
+            uris.internal.getPipeline.apply(Reify.viewInGlobalWindow(view, uriCoder))
           }
-        case shardCount =>
-          val uri = ShardedSparkeyUri(basePath, self.context.options)
-          require(!uri.exists, s"Sparkey URI $uri already exists")
-          logger.info(s"Saving as Sparkey with $shardCount shards: $uri")
-
-          self.transform { collection =>
-            collection
-              .groupBy { case (k, _) => floorMod(w.shardHash(k), shardCount).toShort }
-              .map {
-                case (shard, xs) =>
-                  writeToSparkey(
-                    uri.sparkeyUriForShard(shard, shardCount),
-                    maxMemoryUsage,
-                    xs.asJava
-                  )
-              }
-              .groupBy(_ => ())
-              .map(_ => uri: SparkeyUri)
+          .map { _ =>
+            if (numShards <= 1) {
+              val src = FileSystems
+                .`match`(basePath + "/*")
+                .metadata()
+                .asScala
+                .map(_.resourceId())
+                .sortWith(_.getFilename < _.getFilename)
+                .asJava
+              val dst = SparkeyUri.extensions
+                .map(ext => FileSystems.matchNewResource(s"$basePath$ext", false))
+                .asJava
+
+              FileSystems.rename(src, dst)
+
+              nonShardedUri
+            } else {
```

Comment:
this is somewhat ugly but needed to keep the filename's backwards compat, important?

basically renaming `gs://example/path/part-0000-of-0001.{spi, spl}` to `gs://example/path.{spi, spl}`

Reply:
yeah that's what I figured let's keep this for now.